File: airbyte-integrations/bases/bases-destination-jdbc-async/src/main/java/io/airbyte/integrations/destination/jdbc/copy/StreamCopierFactory.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.jdbc.copy;

import io.airbyte.db.jdbc.JdbcDatabase;
import io.airbyte.integrations.destination.StandardNameTransformer;
import io.airbyte.integrations.destination.jdbc.SqlOperations;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream;

public interface StreamCopierFactory<T> {

  StreamCopier create(String configuredSchema,
                      T config,
                      String stagingFolder,
                      ConfiguredAirbyteStream configuredStream,
                      StandardNameTransformer nameTransformer,
                      JdbcDatabase db,
                      SqlOperations sqlOperations);

  static String getSchema(final String namespace, final String configuredSchema, final StandardNameTransformer nameTransformer) {
    if (namespace != null) {
      return nameTransformer.convertStreamName(namespace);
    } else {
      return nameTransformer.convertStreamName(configuredSchema);
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/copy/StreamCopierFactory.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.jdbc.copy

import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.destination.StandardNameTransformer
import io.airbyte.cdk.integrations.destination.jdbc.SqlOperations
import io.airbyte.protocol.models.v0.ConfiguredAirbyteStream

interface StreamCopierFactory<T> {
    fun create(
        configuredSchema: String?,
        config: T,
        stagingFolder: String?,
        configuredStream: ConfiguredAirbyteStream?,
        nameTransformer: StandardNameTransformer?,
        db: JdbcDatabase?,
        sqlOperations: SqlOperations?
    ): StreamCopier

    companion object {
        @JvmStatic
        fun getSchema(
            namespace: String?,
            configuredSchema: String,
            nameTransformer: StandardNameTransformer
        ): String {
            return if (namespace != null) {
                nameTransformer.convertStreamName(namespace)
            } else {
                nameTransformer.convertStreamName(configuredSchema)
            }
        }
    }
}
